facilitator/src/http.rs

```diff
@@ -1,65 +1,109 @@
-use anyhow::{anyhow, Context, Result};
-use ureq::{Request, Response, SerdeValue};
+use anyhow::{Context, Result};
+use lazy_static::lazy_static;
+use std::time::Duration;
+use ureq::{Agent, AgentBuilder, Error, Request, Response, SerdeValue};
+use url::{ParseError, Url};
 
 use crate::gcp_oauth::OauthTokenProvider;
+lazy_static! {
+    pub(crate) static ref HTTP: Http = Http::new();
+    static ref EXAMPLE_URL: Url =
+        Url::parse("https://example.com").expect("example url did not parse");
+}
+pub(crate) struct Http {
+    agent: Agent,
+}
+
+impl Http {
+    pub fn new() -> Self {
+        Http {
+            agent: AgentBuilder::new().timeout(Duration::from_secs(10)).build(),
+        }
+    }
+
+    pub(crate) fn prepare_request(&self, mut parameters: RequestParameters) -> Request {
+        let request = self.agent.request_url(parameters.method, parameters.url);
+        parameters.get_authenticated_request(request)
+    }
+}
 
 /// Struct containing parameters for send_json_request
-#[derive(Debug, Default)]
-pub(crate) struct JsonRequestParameters<'a> {
-    /// The ureq::Request to be sent
-    pub request: Request,
+#[derive(Debug)]
+pub(crate) struct RequestParameters<'a> {
+    /// The url to request
+    pub url: &'a Url,
+    /// The method of the request (GET, POST, etc)
+    pub method: &'a str,
     /// If this field is set, the request with be sent with an "Authorization"
     /// header containing a bearer token obtained from the OauthTokenProvider.
-    /// If unset, the request is sent unauthenticated.
+    /// If unset, the request is sent unauthenticated if the token field is also not set.
     pub token_provider: Option<&'a mut OauthTokenProvider>,
-    /// Timeout in milliseconds for connections. If unset, a default timeout of
-    /// 10 seconds is applied.
-    pub connect_timeout_millis: Option<u64>,
-    /// Timeout in milliseconds for reads. If unset, a default timeout of 10
-    /// seconds is applied.
-    pub read_timeout_millis: Option<u64>,
-    /// The JSON body to be sent in the request, constructed with ureq::json!
-    pub body: SerdeValue,
+    /// If the token_provider field isn't set, and this field is set, the request will
+    /// be sent with an "Authorization" header containing a bearer token obtained from
+    /// the OauthTokenProvider.
+    pub token: Option<&'a str>,
```

Comment:
I think this is acceptable as is. However, having separate fields for `token_provider` and `token` is unfortunate, because now it's possible to provide both a `token_provider` and a `token`, but that's clearly invalid. I know your implementation will simply prefer the `token_provider`, but the principle of "make invalid states unrepresentable" suggests that we could find a better way. I have two different suggestions for improving this (in both cases you remove the `token` field on `RequestParameters`:

- you could define a new trait `OauthTokenProvider` that defines the method `ensure_oauth_token`, then refactor the existing `OauthTokenProvider` into an implementation of that trait and rename it `GcpOauthTokenProvider`, and then provide a struct like `StaticOauthTokenProvider` which would just wrap a `String` and return it from its implementation of `ensure_oauth_token`. Then the `token_provider` member of `RequestParameters` would be `Option<dyn OauthTokenProvider>` (or `Option<Box<dyn OauthTokenProvider>>`, whichever appeases the compiler) and you would always call `ensure_oauth_token` from `get_authenticated_request`. Callers would be responsible for instantiating a `StaticOauthTokenProvider` if they happen to have a token in hand already (or even better, you might be able to provide a [`std::convert::From<String>`](https://doc.rust-lang.org/std/convert/trait.From.html) implementation on `StaticOauthTokenProvider` so that they can pass their `String` and it will  automatically become an `OauthTokenProvider`!
- Make the `token_provider` field of `RequestParameters` into an [`FnMut`](https://doc.rust-lang.org/std/ops/trait.FnMut.html) that returns the token as a `String`. In the case where the caller has an `OauthTokenProvider`, they would provide a closure that calls `ensure_oauth_token()`. If the caller has an Oauth token as a string, they provide a closure that just returns that string.

Like I said, the current approach is fine, but I think either of the approaches above are a teeny bit better, and interesting Rust exercises.

Reply:
Thank you for this detailed explanation! The first option felt more rusty :rabbit:, so I went with that.